### PR TITLE
Delete API Gateway if cluster up fails

### PR DESCRIPTION
**Pull request by deliahu**
---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)

### CODE REVIEW

File: cli/cmd/cluster.go [hunk 1]
```diff
@@ -768,10 +779,38 @@ func createDashboard(awsClient *aws.Client, dashboardName string) error {
...
 		return err
 	}
 
 	fmt.Println(" ✓")
 
 	return nil
 }
+
+// createOrClearAPIGateway creates an API gateway for the cluster (or clears an existing one if it already exists)
+func createOrClearAPIGateway(awsClient *aws.Client, clusterName string) error {
```
Comment (tthebst) [inline cli/cmd/cluster.go:792]: maybe change function name because it clears AND creates the gateway

File: cli/cmd/cluster.go [hunk 2]
```diff
@@ -753,8 +764,8 @@ func CreateLogGroupIfNotFound(awsClient *aws.Client, logGroup string) error {
 	return nil
 }
 
-// createDashboard creates a new dashboard (or clears an existing one if it already exists)
-func createDashboard(awsClient *aws.Client, dashboardName string) error {
+// createOrClearDashboard creates a new dashboard (or clears an existing one if it already exists)
+func createOrClearDashboard(awsClient *aws.Client, dashboardName string) error {
```
Comment (tthebst) [inline cli/cmd/cluster.go:768]: maybe change function name because it clears AND creates the dashboard

File: cli/cmd/cluster.go [hunk 3]
```diff
@@ -768,10 +779,38 @@ func createDashboard(awsClient *aws.Client, dashboardName string) error {
...
+		return errors.Append(err, fmt.Sprintf("\n\nunable to delete existing vpc link with tag %s=%s; please delete it via the api gateway console: https://%s.console.aws.amazon.com/apigateway/main/vpc-links", clusterconfig.ClusterNameTag, clusterName, awsClient.Region))
+	}
+
+	_, err = awsClient.DeleteAPIGatewayByTag(clusterconfig.ClusterNameTag, clusterName)
+	if err != nil {
+		fmt.Print("\n\n")
+		return errors.Append(err, fmt.Sprintf("\n\nunable to delete existing api gateway with tag %s=%s; please delete it via the api gateway console: https://%s.console.aws.amazon.com/apigateway/main/apis", clusterconfig.ClusterNameTag, clusterName, awsClient.Region))
+	}
+
+	tags := map[string]string{clusterconfig.ClusterNameTag: clusterName}
```
[the remark this answers is not in the document]
Comment (deliahu) [reply cli/cmd/cluster.go:807]: Yes absolutely, good catch!

File: cli/cmd/cluster.go (hunk 1 above)
Comment (deliahu) [reply cli/cmd/cluster.go:792]: I renamed it to `createOrReplaceAPIGateway`, does that sounds ok to you? My thinking is it creates it if it doesn't exist, and it replaces it with a new one if it did exist.

File: cli/cmd/cluster.go (hunk 2 above)
Comment (deliahu) [reply cli/cmd/cluster.go:768]: My thinking was that it creates it if it doesn't exist, or it clears it if it did exist. Do you think that makes sense, or you still suggest changing the name?

File: cli/cmd/cluster.go (hunk 1 above)
Comment (tthebst) [reply cli/cmd/cluster.go]: this is better

File: cli/cmd/cluster.go (hunk 2 above)
Comment (tthebst) [reply cli/cmd/cluster.go:768]: with dashboard clear makes sense

File: cli/cmd/cluster.go [hunk 4]
```diff
@@ -156,16 +156,25 @@ var _upCmd = &cobra.Command{
...
+		}
+
+		err = createOrReplaceAPIGateway(awsClient, clusterConfig.ClusterName, clusterConfig.Tags)
 		if err != nil {
 			exit.Error(err)
 		}
 
 		out, exitCode, err := runManagerUpdateCommand("/root/install.sh", clusterConfig, awsCreds, _flagClusterEnv)
 		if err != nil {
+			_, _ = awsClient.DeleteAPIGatewayByTag(clusterconfig.ClusterNameTag, clusterConfig.ClusterName) // best effort deletion
```
Comment (vishalbollu) [inline cli/cmd/cluster.go:171]: `_, _ = ` can be removed

File: manager/get_api_gateway_endpoint.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import boto3
+import os
+
+
+def get_api_gateway_endpoint():
+    cluster_name = os.environ["CORTEX_CLUSTER_NAME"]
```
Comment (vishalbollu) [inline manager/get_api_gateway_endpoint.py:20]: Since there are only 2 variables, it may be easier to follow the code that calls this python file if these values are accepted as args

File: manager/get_api_gateway_id.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import boto3
+import os
+
+
+def get_api_gateway_id():
+    cluster_name = os.environ["CORTEX_CLUSTER_NAME"]
```
Comment (vishalbollu) [inline manager/get_api_gateway_id.py:20]: Since there are only 2 variables, it may be easier to follow the code that calls this python file if these values are accepted as args

File: pkg/lib/aws/apigateway.go [hunk 7]
```diff
@@ -25,6 +25,34 @@ import (
...
+	}
+
+	_, err = c.APIGatewayV2().CreateStage(&apigatewayv2.CreateStageInput{
+		ApiId:      createAPIResponse.ApiId,
+		AutoDeploy: aws.Bool(true),
+		StageName:  aws.String("$default"),
+		Tags:       aws.StringMap(tags),
+	})
+	if err != nil {
+		_ = c.DeleteAPIGateway(*createAPIResponse.ApiId) // best effort cleanup
```
Comment (vishalbollu) [inline pkg/lib/aws/apigateway.go:49]: `_ =` can be dropped here

File: manager/get_api_gateway_endpoint.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    client_apigateway = boto3.client("apigatewayv2", region_name=region)
+
+    paginator = client_apigateway.get_paginator("get_apis")
+    for api_gateway_page in paginator.paginate():
+        for api_gateway in api_gateway_page["Items"]:
+            if api_gateway["Tags"].get("cortex.dev/cluster-name") == cluster_name:
+                return api_gateway["ApiEndpoint"]
+
+    raise Exception(
+        f"Could not find api gateway with tag cortex.dev/cluster-name={cluster_name} in {region}"
```
Comment (vishalbollu) [inline manager/get_api_gateway_endpoint.py:31]: `Could not find api gateway with tag` -> `Could not get endpoint for api gateway with tag`

File: manager/get_api_gateway_endpoint.py (hunk 5 above)
Comment (deliahu) [reply manager/get_api_gateway_endpoint.py:20]: The other python files, e.g. `create_gateway_integration.py`, also pull from the environment variables, so I think for now I lean towards leaving it as is. I'm actually leaning toward moving more of the api gateway code from `install.sh` to python, so at that point we can convert to only using args, does that sound ok?

File: manager/get_api_gateway_id.py (hunk 6 above)
Comment (deliahu) [reply manager/get_api_gateway_id.py:20]: (see comment above)

File: manager/get_api_gateway_endpoint.py (hunk 8 above)
Comment (deliahu) [reply manager/get_api_gateway_endpoint.py:31]: I think the current error message is a little more descriptive/accurate, since if this exception is raised, it means that there was not API gateway that had the tag. I'll reword it to make it more clear: "your cluster's api gateway (with tag cortex.dev/cluster-name={cluster_name}) does not exist". Does that sound ok?
